solidity/src/CadenceRandomConsumer.sol

```diff
@@ -69,27 +193,63 @@ abstract contract CadenceRandomConsumer is CadenceArchWrapper {
-        uint32 requestIndex = requestId - 1;
+        uint256 requestIndex = requestId - 1;
         require(requestIndex < _requests.length, "Invalid request ID - value exceeds the number of existing requests");
-        Request memory request = _requests[requestIndex];
 
-        // Ensure that the request is fulfilled at a Flow block height greater than the one at which the request was made
-        uint64 flowHeight = _flowBlockHeight();
-        require(request.flowHeight < flowHeight, "Cannot fulfill request until subsequent Flow network block height");
+        // Access & validate the request
+        Request storage request = _requests[requestIndex];
+        _validateRequest(request);
+        request.fulfilled = true; // Mark the request as fulfilled
 
-        // Get the random source for the Flow block at which the request was made
-        uint64 randomResult = _getRandomSource(request.flowHeight); // returns bytes32
-        // Pack the randomResult into a uint64, hashing with the requestId to vary results across shared block heights.
-        // The random seed returned from Cadence Arch is only 32 bytes. Here only 8 bytes were required, but if an
-        // implementing contract requires more random bytes, the source should be expanded into any arbitrary number of
-        // bytes using a PRG.
-        randomResult = uint64(uint256(keccak256(abi.encodePacked(randomResult, requestId))));
+        // Get the random source for the Flow block at which the request was made, emit & return
+        bytes32 randomSource = _getRandomSource(request.flowHeight);
 
-        emit RandomnessFulfilled(requestId, request.flowHeight, request.evmHeight, randomResult);
+        emit RandomnessSourced(requestId, request.flowHeight, request.evmHeight, randomSource);
 
-        // Return the random result
-        return randomResult;
+        return randomSource;
+    }
+
+    /**
+     * @dev This method aggregates 256 bits of randomness by calling _revertibleRandom() 4 times.
+     *
+     * @return randomValue The aggregated 256 bits of randomness.
+     */
+    function _aggregateRevertibleRandom256() private view returns (uint256) {
+        // Call _revertibleRandom() 4 times to aggregate 256 bits of randomness
+        uint256 randomValue = uint256(_revertibleRandom());
+        randomValue |= (uint256(_revertibleRandom()) << 64);
+        randomValue |= (uint256(_revertibleRandom()) << 128);
+        randomValue |= (uint256(_revertibleRandom()) << 192);
+        return randomValue;
+    }
+
+    /**
+     * @dev This method returns the most significant bit of a uint64.
+     *
+     * @param x The input value.
+     * @return bits The most significant bit of the input value.
+     */
+    function _mostSignificantBit(uint64 x) private pure returns (uint64) {
+        uint64 bits = 0;
+        while (x > 0) {
+            x >>= 1;
+            bits++;
+        }
+        return bits;
+    }
```

Comment:
following up on my last comment, this can become:
```suggestion
    function _getMask(uint64 x) private pure returns (uint64) {
        uint64 mask = 0;
        while (mask & x != x) {
            mask = (mask<<1) | 1;
        }
        return mask;
    }
```